### PR TITLE
Compiler B: Built-in libraries

### PR DESCRIPTION
**Pull request by suli-97**

### CODE REVIEW

File: designs/Compiler B: Built-in libraries-design.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+Since we will have to high-level functions, we suggest to add another basic type function like below in ast.ts
+
+Type = 
+
+​	|... 
+
+​	| {
+
+​		tag: "function",
```
Comment (KamranAlipour) [inline designs/Compiler B: Built-in libraries-design.md:110]: Interesting idea, why define function as a new type? are you planning to parse function as a type in general ( e.g. class<function>, etc.)?

File: designs/Compiler B: Built-in libraries-design.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+gcd :: int -> int -> int
+
+comb :: int -> int -> int
+
+perm :: int -> int -> int
+
+# Value representation and memory layout
+
+we think a better way to represent functions is to place them all, including methods and functions, into the v table. And a function     is just a int32 variable in wasm indicating the its index in vtable. In this way, we can treat functions just like any other variables, so we can write high-level functions like map and reduce.
```
Comment (KamranAlipour) [inline designs/Compiler B: Built-in libraries-design.md:144]: Are you planning to have separate memory for your built-in functions? If not, how will your functions be indexed alongside the user-defined functions in the vtable?

**Review by KamranAlipour — COMMENTED** [review]
Besides the functions proposed, you can look at supporting multiple parameters for `print()`. Also, think of type conversions (e.g. `int()`,`bool()`)

File: designs/Compiler B: Built-in libraries-design.md (hunk 1 above)
Comment (suli-97) [reply designs/Compiler B: Built-in libraries-design.md:110]: I give up this idea.

File: designs/Compiler B: Built-in libraries-design.md (hunk 2 above)
Comment (suli-97) [reply designs/Compiler B: Built-in libraries-design.md:144]: I give up this idea

**Pull request closed without merging**